lib/util/result.ts

```diff
@@ -1,73 +1,441 @@
+/* eslint-disable promise/no-nesting */
+import { logger } from '../logger';
+
 interface Ok<T> {
-  readonly success: true;
-  readonly value: T;
+  readonly ok: true;
+  readonly val: NonNullable<T>;
+  readonly err?: never;
 }
 
 interface Err<E> {
-  readonly success: false;
-  readonly error: E;
+  readonly ok: false;
+  readonly err: NonNullable<E>;
+  readonly val?: never;
 }
 
 type Res<T, E> = Ok<T> | Err<E>;
 
 export class Result<T, E = Error> {
-  static ok<T>(value: T): Result<T, never> {
-    return new Result({ success: true, value });
+  private constructor(private readonly res: Res<T, E>) {}
+
+  static ok<T>(val: NonNullable<T>): Result<T, never> {
+    return new Result({ ok: true, val });
   }
 
-  static err(): Result<never, true>;
-  static err<E>(e: E): Result<never, E>;
-  static err<E>(e?: E): Result<never, E> | Result<never, true> {
-    if (typeof e === 'undefined' && arguments.length === 0) {
-      return new Result({ success: false, error: true });
+  static err<E>(err: NonNullable<E>): Result<never, E> {
+    return new Result({ ok: false, err });
+  }
+
+  /**
+   * Wrap a callback or promise in a Result in such a way that any thrown errors
+   * are caught and wrapped with `Result.err()`.
+   *
+   * In case of a promise, the `AsyncResult` is returned.
+   * Use `.unwrap()` to get the `Promise<Result<T, E>>` from `AsyncResult`.
+   *
+   *   ```ts
+   *
+   *   // SYNC
+   *   const parse = (json: string) => Result.wrap(() => JSON.parse(json));
+   *
+   *   const { val, err } = parse('{"foo": "bar"}').unwrap();
+   *   expect(val).toEqual({ foo: 'bar' });
+   *   expect(err).toBeUndefined();
+   *
+   *   const { val, err } = parse('!!!').unwrap();
+   *   expect(val).toBeUndefined();
+   *   expect(err).toBeInstanceOf(SyntaxError);
+   *
+   *   // ASYNC
+   *   const request = (url: string) => Result.wrap(http.get(url));
+   *
+   *   const { val, err } = await request('https://example.com').unwrap();
+   *   expect(val).toBeString();
+   *   expect(err).toBeUndefined();
+   *
+   *   ```
+   */
+  static wrap<T, E = Error>(callback: () => NonNullable<T>): Result<T, E>;
+  static wrap<T, E = Error>(promise: Promise<Result<T, E>>): AsyncResult<T, E>;
+  static wrap<T, E = Error>(
+    promise: Promise<NonNullable<T>>
+  ): AsyncResult<T, E>;
+  static wrap<T, E = Error>(
+    input:
+      | (() => NonNullable<T>)
+      | Promise<Result<T, E>>
+      | Promise<NonNullable<T>>
+  ): Result<T, E> | AsyncResult<T, E> {
+    if (input instanceof Promise) {
+      return AsyncResult.wrap(input as never);
     }
 
-    const error = e as E;
-    return new Result({ success: false, error });
+    try {
+      const result = input();
+      return Result.ok(result);
```

Comment:
we should also support async callback, so check here for promise again?

Reply:
There is no much sense for callback returning `Promise`. For synchronous code, callback is being used because there is no way to wrap it to the `try-catch` block. If caller side wants to use asynchronous callback, it's simple to get the promise value first, then wrap it with this function.